dissect/cstruct/cstruct.py

```diff
@@ -280,37 +314,143 @@ def resolve(self, name: str) -> BaseType:
 
         raise ResolveError(f"Recursion limit exceeded while resolving type {name}")
 
+    def _make_type(
+        self,
+        name: str,
+        bases: Iterator[object],
+        size: Optional[int],
+        *,
+        alignment: Optional[int] = None,
+        attrs: dict[str, Any] = None,
+    ) -> type[BaseType]:
+        """Create a new type class bound to this cstruct instance.
+
+        All types are created using this method. This method automatically binds the type to this cstruct instance.
+        """
+        attrs = attrs or {}
+        attrs.update(
+            {
+                "cs": self,
+                "size": size,
+                "dynamic": size is None,
+                "alignment": alignment or size,
+            }
+        )
+        return types.new_class(name, bases, {}, lambda ns: ns.update(attrs))
+
+    def _make_array(self, type_: MetaType, num_entries: Optional[int | Expression]) -> ArrayMetaType:
+        null_terminated = num_entries is None
+        dynamic = isinstance(num_entries, Expression) or type_.dynamic
+        size = None if (null_terminated or dynamic) else (num_entries * type_.size)
+        name = f"{type_.__name__}[]" if null_terminated else f"{type_.__name__}[{num_entries}]"
+
+        bases = (type_.ArrayType,)
+
+        attrs = {
+            "type": type_,
+            "num_entries": num_entries,
+            "null_terminated": null_terminated,
+            "dynamic": dynamic,
```

Comment:
Left it out here. I don't think it makes much sense to explicitly check for `dynamic` in `_make_type()` since it shouldn't be possible to create a type where `size != None && dynamic`.